### PR TITLE
Version 8.7 of #6328 (fix #6313: lost goals in nested ltac in refine)

### LINKED ISSUE #6313: ltac:(shelve) does not interact well with pattern matching and (simple) refine
#### Version

8.7.0

#### Description of the problem

This code should work.  It does not:
```coq
Inductive foo := a | b | c.
Goal foo -> foo.
  intro x.
  simple refine (match x with
                 | a => _
                 | b => ltac:(shelve)
                 | c => _
                 end); [ | ].
  (* Error: Tactic failure: Incorrect number of goals (expected 1 tactic). *)
```

Interestingly, this is also a case where `Unshelve` fails to unshelve evars:
```coq
Inductive foo := a | b | c.
Goal foo -> foo.
  intro x.
  unshelve refine (match x with
                   | a as x' => ?[a]
                   | b as x' => ltac:(shelve)
                   | c as x' => ?[b]
                   end).
  exact c.
  Unshelve.
  Fail idtac. (* Error: No such goal. *)
  Grab Existential Variables.
  exact b.
  exact a.
  Show Proof.
  (* (fun x : foo => match x with
                | a => a
                | b => b
                | c => c
                end) *)
```

### PR DESCRIPTION
**Pull request by herbelin**
A version of #6328 for v8.7 which does not change the API.

It still does not take into account the shelved/given-up status of remaining goals in `ltac:()` (as I plan to do in #6328) but it fixes the critical lost of the status of goals which resulted in wrongly shelving all the goals generated up to the point of a `ltac:()` inside a `refine`.

### CODE REVIEW

**Comment by Zimmi48** [comment]
I still don't know whether people get e-mail when they are requested reviews so @ppedrot here is a comment to make sure you see the request for review on this one. (Should be pretty trivial to review.)

**Review by ppedrot — APPROVED** [review]
As far as I'm concerned, this looks correct, but I wouldn't bet on it either. This shelf mess is a real nightmare and I'm pretty sure there is a corner case that will make this explode. As long as all our tests compile, this should be OK to merge though.

**Comment by Zimmi48** [comment]
Note that this is a PR on v8.7. Do you think this is safe to merge just before the 8.7.1 release? Or do you think it is a dangerous area and it would be safer to fix this bug only in master?

**Comment by ppedrot** [comment]
It doesn't seem so critical as nobody realized this until recently. But this would also make me think that if the fix goes in, nobody would notice either. I think that passing the Travis tests gives a good idea of whether this is actually sensible.

**Comment by Zimmi48** [comment]
OK, all this uncertainty leaves me a bad feeling so I'm going to postpone this (merging just before a release means merging without really testing). I'll merge the PR after #6328 has been merged in master and tested for a few weeks... Thus, if no issue is detected in the meantime, this should land in 8.7.2.

**Comment by herbelin** [comment]
To clarify, I used "critical" in the sense that it completely disturbs the proof state. It is a not blocking in practice in the sense that it was reported only recently (though I'm suprised that noone fell on it yet: I'm more enclined to think that people wondered whether it was a correct behavior or not, that they could not decide and did not report).

**Comment by JasonGross** [comment]
> though I'm suprised that noone fell on it yet: I'm more enclined to think that people wondered whether it was a correct behavior or not, that they could not decide and did not report

I probably encountered it 5-6 times before I actually reported it; it took me that long to discover that it was `shelve` in `ltac:(...)` that was causing some of my shelf to end up needing `Grab Existential Variables`, rather than some other cause.

**Comment by Zimmi48** [comment]
The main version of this PR has been waiting for activity for a long time so this is too late for this fix to make it into 8.7.2. And anyways, this is a dangerous thing to merge into a stable branch given how it changes the behavior of some Ltac. So closing :/

**Pull request closed without merging**
